### ISSUE TITLE
File enumeration order in directory-scanning mode should be deterministic

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#8776]
Repro: Set up `a.ts`, `b.ts`, each with some statements, and a `tsconfig.json` in a folder using `outFile` then compile.

Result: Depending on which order the OS returns the files (which could change per machine), you might get one `a.ts`'s contents before `b.ts`, or after

We should `.sort()` https://github.com/Microsoft/TypeScript/blob/adbf8fdee0c9f6ba46d106c4ca972e438511cbcb/src/compiler/commandLineParser.ts#L751 here so that TS builds are deterministic

**Comment by basarat** [comment]
Not blindly perhaps.  tsconfig.json `files` order is preserved (https://github.com/Microsoft/TypeScript/issues/2472) and if not using `tsconfig.json` then `tsc _references.ts` order is preserved :rose:

**Comment by RyanCavanaugh** [comment]
Yep, this should only be when no other ordering data is present

**Comment by rotemdan** [comment]
I started using the glob system through the `include` property in `tsconfig.json`. I'm surprised that the globs are not evaluated in written order, and subsequently the output file is not concatenated in the order they are stated (after all, it is an ordered array) as that is necessary for my project's build process. 

I had to resort to a rather 'ugly' solution where I number the directories through a prefix:
```js
"compilerOptions": {
  "outFile": "MyApp.js"
}

"include": [
 "01_Extensions",
 "02_Dependencies",
 "03_Library",
 "04_Program",
 ...
]
```
It seems to work, at least, but doesn't look very professional or like a permanent solution. It would be great if the globs would also provide ordering for the build process and resulting concatenated file. Thanks!
